File: google-sheets-service/services/sheets.js
const { google } = require("googleapis");
const fs = require("fs");
const { OAuth2Client } = require("google-auth-library");
const CLIENT_ID = process.env.CLIENT_ID;
const CLIENT_SECRET = process.env.CLIENT_SECRET;
const REDIRECT_URI = process.env.REDIRECT_URI;
// console.log(CLIENT_ID, CLIENT_SECRET, REDIRECT_URI)
const oAuth2Client = new OAuth2Client(CLIENT_ID, CLIENT_SECRET, REDIRECT_URI);
// const content = fs.readFileSync("token.json", "utf8");
// const credentials = JSON.parse(content);
// let access_token = credentials.access_token;
// let refresh_token = credentials.refresh_token;

// let access_token = process.env.ACCESS_TOKEN;
let refresh_token = process.env.REFRESH_TOKEN;



const mongoose = require("mongoose");
const Sheet = require("../models/sheets");
const Response = require("../models/response");
const { sheets } = require("googleapis/build/src/apis/sheets");

exports.sheets_create_and_add = async (data) => {
    try {
        // const access_token = AUTH_TOKEN || req.session.tokens.access_token;
        try {
            await oAuth2Client.setCredentials({
                access_token: access_token,
            });
        } catch (err) {
            console.log('Refresh token')
            const { tokens } = await oAuth2Client.refreshToken(refresh_token);
            // console.log(tokens);
            await oAuth2Client.setCredentials(tokens);
        }

        // console.log('Got data: ', data);
        let response = data.createdResponse._id;
        let form = data.createdResponse.form._id;
        let title = data.createdResponse.form.title;
        let user = data.createdResponse.user;
        let createdAnswer = data.createdAnswers;
        let answers_array = [];
        createdAnswer.forEach((answer) => {
            let temp = {};
            temp.question_id = answer.question._id;
            temp.question_text = answer.question.text;
            temp.answer_text = answer.text;
            answers_array.push(temp);
        });
        // console.log(answers_array);

        await insert_response(form, response, answers_array);
        await Sheet.findOne({ form: form })
            .exec()
            .then(async (finder) => {
                try {
                    if (finder) {
                        console.log("Sheet exists");
                        let sheet_id = finder.sheet;
                        await add_data_sheet(
                            sheet_id,
                            data.createdResponse.form,
                            answers_array
                        );
                    } else {
                        const spreadsheetTitle = title + "_" + form.slice(-5);
                        let sheet_id = await add_form_sheet(spreadsheetTitle, form);
                        await add_data_sheet(
                            sheet_id,
                            data.createdResponse.form,
                            answers_array
                        );
                    }
                } catch (error) {
                    if (error) console.log('Sheet insertion error: ',error);
                }
            })
            .catch((error) => {
                console.log(error);
            });
    } catch (err) {
        console.log(err);
    }
};

async function add_data_sheet(sheet_id, form, data) {
    try {
        const sheets = google.sheets({ version: "v4", auth: oAuth2Client });
        console.log(data);
        console.log("Adding data");

        let values = [];
        data.forEach((value) => {
            values.push(value.answer_text);
        });

        rows = [values];

        const resource = {
            values: rows,
        };
        const spreadsheetId = sheet_id;
        const valueInputOption = "RAW";
        const range = "Sheet1";

        try {
            const result = await sheets.spreadsheets.values.append({
                spreadsheetId,
                range,
                valueInputOption,
                resource,
            });
            console.log("Data inserted successfully into spreadsheet");
        } catch (err) {
            throw err;
        }
    } catch (err) {
        throw err;
    }
}

async function add_form_sheet(form_title, form_id) {
    const sheets = google.sheets({ version: "v4", auth: oAuth2Client });
    const resource = {
        properties: {
            title: form_title,
        },
    };
    const spreadsheet = await sheets.spreadsheets.create({
        resource,
        fields: "spreadsheetId",
    });
    console.log(`Spreadsheet ID: ${spreadsheet.data.spreadsheetId}`);
    let sheet = new Sheet({
        _id: new mongoose.Types.ObjectId(),
        form: form_id,
        sheet: spreadsheet.data.spreadsheetId,
    });

    sheet
        .save()
        .then(async (res) => {
            return sheet.sheet;
        })
        .catch((err) => {
            throw err;
        });
}

async function insert_response(form, response, answers_array) {
    try {
        // await Response.findOne({ _id: response })
        //     .then((f) => {
        //         if (f) return;
        //     })
        //     .catch((err) => {
        //         throw err;
        //     });
        answers = [];
        answers_array.forEach((answer) => {
            res = new Response({
                response: response,
                form: form,
                question: answer.question_id,
                answer: answer.answer_text,
            });
            answers.push(res);
        });
        Response.insertMany(answers)
            .then((res) => {
                console.log("Answers inserted successfully into the database");
            })
            .catch((err) => {
                throw err;
            });
    } catch (err) {
        throw err;
    }
}
// exports.sheets_write = async (req,res,next)=>{
//   const id = req.params.formId;

// }

File: google-sheets-service/service/sheets.py
import os
from dotenv import load_dotenv
load_dotenv()

db = None

from datetime import datetime
from bson import ObjectId
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from googleapiclient.discovery import build


# Google API Scopes required for sending emails
SCOPES = ['https://www.googleapis.com/auth/spreadsheets']


def load_credentials():
    creds = None
    creds = Credentials(
        token_uri="https://accounts.google.com/o/oauth2/token",
        client_id=os.getenv('CLIENT_ID'),
        client_secret=os.getenv('CLIENT_SECRET'),
        refresh_token=os.getenv('REFRESH_TOKEN'),
        token=os.getenv('ACCESS_TOKEN'),
    ) 
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        # else:
        #     flow = InstalledAppFlow.from_client_secrets_file(credentials_path, SCOPES)
        #     creds = flow.run_local_server(port=0)

        # with open(token_path, 'w') as token:
        #     token.write(creds.to_json())

    return creds


def process_data(data,mongo_client):
    try:
        global db 
        db = mongo_client['test']
        response = data["message"]["createdResponse"]["_id"]
        form = data["message"]["createdResponse"]["form"]["_id"]
        title = data["message"]["createdResponse"]["form"]["title"]
        user =  data["message"]["createdResponse"]["user"]
        createdAnswer = data["message"]["createdAnswers"]
        answers_array = []
        for answer in createdAnswer:
            temp = {}
            temp["question_id"] = answer["question"]["_id"]
            temp["question_text"] = answer["question"]["text"]
            temp["answer_text"] = answer["text"];  
            answers_array.append(temp)          
        
        insert_response(form, response, answers_array)
        add_data_to_sheet(form,answers_array)

    except Exception as e:
        print('Error adding data to sheet:', e)

def insert_response(form, response, answers_array):
    try:        
        answers = []
        for answer in answers_array:
            res = {
                "response": ObjectId(response),
                "form": ObjectId(form),
                "question": ObjectId(answer["question_id"]), 
                "text": answer["answer_text"],
                "createdAt": datetime.today().replace(microsecond=0)
            }
            answers.append(res)
        db.responses.insert_many(answers)
        print("Answers inserted successfully into the database")

    except Exception as e:
        print(f"Error inserting responses into the database: {e}")
    


def add_data_to_sheet(form, answers):
    try:
        values = []
        for value in answers:
            values.append(value['answer_text'])
        rows = values

        resource = {
            'values': [rows]
        }
        
        sheet_id = get_form_sheet(form)

        spreadsheet_id = sheet_id
        range_ = 'Sheet1'
        sheets = build('sheets', 'v4', credentials=load_credentials())
        response = sheets.spreadsheets().values().append(
            spreadsheetId=spreadsheet_id,
            range=range_,
            valueInputOption='RAW',
            body=resource
        ).execute()

        print('Data inserted successfully into spreadsheet')
    except Exception as e:
        print('Unable to insert data into the database',e)
        raise Exception(e)
        



def get_form_sheet(form_id):
    try:
        existing_entry = db.sheets.find_one({'form': form_id})
        if existing_entry:
            print("Sheet entry already exists in MongoDB.")
            return existing_entry['sheet']  
              
        spreadsheet_title = 'Form' + "_" + form_id[-5:]
        sheets = build('sheets', 'v4', credentials=load_credentials())
        spreadsheet = sheets.spreadsheets().create(
            body={'properties': {'title': spreadsheet_title}}
        ).execute()

        spreadsheet_id = spreadsheet['spreadsheetId']
        print(f"Spreadsheet ID: {spreadsheet_id}")

        # Insert document into MongoDB
        sheet = {
            '_id': ObjectId(),
            'form': form_id,
            'sheet': spreadsheet_id,
            'createdAt': datetime.today().replace(microsecond=0)
        }
        db.sheets.insert_one(sheet)

        print("Sheet information inserted into MongoDB successfully.")

        return spreadsheet_id

    except Exception as e:        
        print(f"Error creating Google Sheet: {e}")
        raise Exception(e)
